### PR TITLE
Add support to integer modulo operator

### PR DESCRIPTION
**Pull request by ebassi**
Having support for the '%' operator makes it easier to implement
even/odd version checks, like:

    enable_debug = get_option('enable-debug')
    if enable_debug == 'auto'
      if minor_version % 2 == 0
        enable_debug = 'minimum'
      else
        enable_debug = 'yes'
      endif
    endif

which would be impossible without resorting to less obvious long-hand
forms like:

  a - (b * (a / b))

### CODE REVIEW

**Comment by nirbheek** [comment]
Perhaps it also makes sense to add `is_even` and `is_odd` methods to integer types for further clarity? `%` would make sense even in the presence of those methods just for completeness.

**Comment by ebassi** [comment]
You mean:

    assert(2.is_even())

a la Ruby, or:

    foo = 2
    assert(foo.is_even())

a la Python?

**Comment by nirbheek** [comment]
Is there a difference in this case implementation-wise? Unlike Python, Meson doesn't special-case the int type (although there aren't any int methods yet). AFAICT it would be identical to how string methods are currently implemented.

**Comment by ebassi** [comment]
Ah, yes; after your comment I went through the implementation — though it doesn't seem to be working. I can only return an integer instead of a boolean value.

**Comment by jpakkane** [comment]
There should also be some tests to ensure precedence works correctly. At least (according to Python):

    2*1%2 == 0
    2+1%2 == 3
    9/9%2 == 1
    9-9%2 == 8

**Comment by nirbheek** [comment]
@jpakkane Any idea why `is_even()` can't return a bool type?

**Comment by jpakkane** [comment]
Weeeeird. It should be able to. I tested returning bools from a string method and it worked just fine.

**Comment by nirbheek** [comment]
Okay, everything is actually working fine. 

```
assert(2.is_even() == true, 'int is_even() broken')
assert(2.is_odd() == false, 'int is_odd() broken')
```

This works fine. The reason why `== 1` also works is because Python 3 evaluates `1` to `True`. Do we want to keep that behaviour?

**Comment by ebassi** [comment]
No idea why — probably some PEBKAC — and now it's working. I've also fixed the precedence for the modulo operator.

File: test cases/common/68 number arithmetic/meson.build [hunk 1]
```diff
@@ -23,6 +23,8 @@ endif
 
 assert((5 % 2) == 1, 'Integer modulo (odd) is broken')
 assert((4 % 2) == 0, 'Integer modulo (even) is broken')
+assert(2.is_even() == 1, 'int is_even() broken')
+assert(2.is_odd() == 0, 'int is_odd() broken')
```
Comment (nirbheek) [inline test cases/common/68 number arithmetic/meson.build]: We should also test for the negative case here:

```
assert(3.is_even() == false, 'int.is_even() is broken for odd values')
assert(5.is_odd() == true, 'int.is_odd() is broken for odd values')
```

The original test will pass if `is_even()` always returns `true` and `is_odd()` always returns `false`.

Comment (ebassi) [inline test cases/common/68 number arithmetic/meson.build]: Good point.
